### PR TITLE
add `sideEffects: false` to package json in esm folder

### PR DESCRIPTION
**Pull request by jeetiss**
this PR is part of 
- #661

---

## PR-Codex overview
This PR focuses on optimizing the `viem` package by marking it as side effects free. Notable changes include:

### Detailed summary
- `viem` package marked as side effects free
- `build:esm` script now includes `sideEffects` property in `package.json`
- Added `size-limit` checks for `viem (cjs)`, `viem (tree-shaking check)`, and `viem/chains`
- Updated `simple-git-hooks` configuration.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by fubhy — APPROVED** [review]

**Comment by fubhy** [comment]
Thanks. You can also introduce the size-limit probes here imho. That way we can easily iterate on them in the follow-up with the annotations.

**Comment by fubhy** [comment]
Thanks @jeetiss! That's good to go as far as I'm concerned. I think we can add even more checks to the size-limit config later on for good measure but that can follow in another PR.

LGTM!

**Review by tmm — COMMENTED** [review]
Want to add a changeset for `package.json#sideEffects`? Then lgtm.

File: .changeset/modern-rockets-fly.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"viem": patch
+---
+
+marked package as side effects free
```
Comment (tmm) [inline .changeset/modern-rockets-fly.md:5]:
```suggestion
Marked package as side effects free
```

**Pull request merged**